docs: show component signal APIs as the standard approach (#58572)
angular/angular: adev/src/content/guide/components/inputs.md

Tip: If you're familiar with other web frameworks, input properties are similar to _props_.

When you use a component, you commonly want to pass some data to it. A component specifies the data that it accepts by declaring
**inputs**:

<docs-code language="ts" highlight="[5]">
import {Component, input} from '@angular/core';

@Component({/*...*/})

> export class CustomSlider { // Declare an input named 'value' with a default value of zero. value = input(0); }

Reviewer: ```suggestion
export class CustomSlider { 
  // Declare an input named 'value' with a default value of zero. 
  value = input(0);
}
```

<custom-slider [value]="50" />
```

If an input has a default value, TypeScript infers the type from the default value:

```typescript
@Component({/*...*/})
export class CustomSlider {
  // TypeScript infers that this input is a number, returning InputSignal<number>.
  value = input(0);
}
```

> You can explicitly declare a type for the input by specifying a generic parameter to the function. If an input is not set, its value is `undefined`.

Reviewer 2: Minor nit: this isn't the case for inputs that have a default value, though that isn't clear from the current language. Maybe:
```suggestion
You can explicitly declare a type for the input by specifying a generic parameter to the function. If a default value for the input is not provided, Angular will use `undefined`.
```

Author: Changed to 
```markdown
If an input without a default value is not set, its value is `undefined`:
```

<custom-slider [value]="50" />
```

If an input has a default value, TypeScript infers the type from the default value:

```typescript
@Component({/*...*/})
export class CustomSlider {
  // TypeScript infers that this input is a number, returning InputSignal<number>.
  value = input(0);
}
```

You can explicitly declare a type for the input by specifying a generic parameter to the function:

```typescript
@Component({/*...*/})
export class CustomSlider {
  // Produces an InputSignal<number | undefined>.
  value = input<number>();
}
```

> If an input is not set, its value is `undefined`.

Reviewer 3: Should this line be moved above the example? I feel its needed to fully understand the example

The `input` function returns an `InputSignal`. You can read the value by calling the signal:

<docs-code language="ts" highlight="[5]">
import {Component, input} from '@angular/core';

@Component({/*...*/})
export class CustomSlider {
  // Declare an input named 'value' with a default value of zero. 
  value = input(0);

  // Create a computed expression that reads the value input
  label = computed(() => `The slider's value is ${this.value()}`); 
}
</docs-code>

Signals created by the `input` function are read-only.

## Required inputs

You can declare that an input is `required` by calling `input.required` instead of `input`:

<docs-code language="ts" highlight="[3]">
@Component({/*...*/})
export class CustomSlider {
  // Declare a required input named value. Returns an `InputSignal<number>`.
  value = input.required<number>();
}
</docs-code>

Angular enforces that required inputs _must_ be set when the component is used in a template. If you try to use a component without specifying all of its required inputs, Angular reports an error at build-time.

Required inputs do not automatically include `undefined` in the generic parameter of the returned `InputSignal`.

> ## Customizing inputs

Reviewer 3: s/customizing/configuring

<docs-code language="ts" highlight="[6]">
@Component({
  selector: 'custom-slider',
  /*...*/
})
export class CustomSlider {

>   label = input('', {transform: () => trimString});

Reviewer 2: As written this wouldn't work. I think this is meant to be:
```suggestion
  label = input('', {transform: trimString});
```

Author: Good catch, fixed

}

> function trimString(value: string | undefined) { return value?.trim() ?? ''; }

Reviewer 2: (annotating the return type makes the intention here clear)
```suggestion
function trimString(value: string | undefined): string { return value?.trim() ?? ''; }
```

Author: Done

**Input transform function must be statically analyzable at build-time.** You cannot set transform functions conditionally or as the result of an expression evaluation.

**Input transform functions should always be [pure functions](https://en.wikipedia.org/wiki/Pure_function).** Relying on state outside the transform function can lead to unpredictable behavior.

#### Type checking

When you specify an input transform, the type of the transform function's parameter determines the types of values that can be set to the input in a template.

<docs-code language="ts">
@Component({/*...*/})
export class CustomSlider {
  widthPx = input('', {transform: appendPx});
}

> function appendPx(value: number) { return `${value}px`; }

Reviewer 2: ```suggestion
function appendPx(value: number): string { return `${value}px`; }
```

Author: Done

While you should generally avoid aliasing inputs for components, this feature can be useful for renaming properties while preserving an alias for the original name or for avoiding collisions with the name of native DOM element properties.

## Model inputs

> **Model inputs** are a special type of input that enable a component to propagate new values back to another component.

Reviewer 2: ```suggestion
**Model inputs** are a special type of input that enable a component to propagate new values back to the parent component.
```

Author: Changed to 
```markdown
**Model inputs** are a special type of input that enable a component to propagate new values back to its parent component.
```

})
export class MediaControls {
  protected volume = 0;
}
```

In the example above, the `CustomSlider` can write values into its `checked` model input, which then propagates those values back to the `isAdmin` property in `UserProfile`. This binding keeps that values of `checked` and `isAdmin` in sync.

### Implicit `change` events

When you declare a model input in a component or directive, Angular automatically creates a corresponding [output](guide/components/outputs) for that model. The output's name is the model input's name suffixed with "Change".

```angular-ts
@Directive({ /* ... */ })
export class CustomCheckbox {
  // This automatically creates an output named "checkedChange".
  // Can be subscribed to using `(checkedChange)="handler()"` in the template.
  checked = model(false);
}
```

Angular emits this change event whenever you write a new value into the model input by calling its `set` or `update` methods.

See [Custom events with outputs](guide/components/outputs) for more details on outputs.

### Customizing model inputs

You can mark a model input as required or provide an alias in the same way as a [standard input](guide/signals/inputs).

> Model inputs do not support input transforms.

Reviewer 2: Is it useful to mention why here? Transforms are one directional.

Author: My thinking is that the doc is already pretty long, and readers don't strictly need to know why to accomplish anything
